### DISCUSSION TITLE
Why does trivy vulnerability scan fails on `openshift-*` namespaces?

### DISCUSSION THREAD

**Discussion by nilesh-akhade** [aquasecurity/trivy#4228] [Q&A]
I am running trivy k8s vulnerability scan on workloads inside `openshift-*` namespaces. It always fails with an error `UNAUTHORIZED: access to the requested resource is not authorized`  .

- This means trivy is unable to inspect the image
  - Because it is not authorized to pull
    - Because the image resides inside a private registry

My questions are: 

- Trivy supports scanning images from [private registries](https://aquasecurity.github.io/trivy/v0.37/docs/advanced/private-registries/). Is it possible to make it work with quay.io? If yes, can we scan these `openshift-*` namespaces?

- Trivy takes 2-3 seconds before it could figure out that the image is not accessible. Are there any retries internally? Can we configure it not to spend much time on such images? From logs, it looks like trivy tries 4 ways to access the image (daemon, podman, contained and GET). Is it possible to skip first 3?

### Following logs are generated when we run the following inside a pod running in the same cluster

> trivy --debug k8s --no-progress --quiet --security-checks=vuln --cache-dir=/tmp --format=json --namespace=openshift-sdn all

```
2023-02-07T15:07:32.216Z        ERROR   Error during vulnerabilities scan: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
        * unable to inspect the image (quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:3ee77e3eb4702fbb1ea5dee009c2318405987c2fab352bdc14c67601515c4b6f): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
        * unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
        * containerd socket not found: /run/containerd/containerd.sock
        * GET https://quay.io/v2/openshift-release-dev/ocp-v4.0-art-dev/manifests/sha256:3ee77e3eb4702fbb1ea5dee009c2318405987c2fab352bdc14c67601515c4b6f: UNAUTHORIZED: access to the requested resource is not authorized; map[]


2023-02-07T15:07:32.216Z        ERROR   Error during vulnerabilities scan: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
        * unable to inspect the image (quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:3e536bf0991bbcddeeb56b91bc8671037592f3fee8b346c775c71e86f846e732): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
        * unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
        * containerd socket not found: /run/containerd/containerd.sock
        * GET https://quay.io/v2/openshift-release-dev/ocp-v4.0-art-dev/manifests/sha256:3e536bf0991bbcddeeb56b91bc8671037592f3fee8b346c775c71e86f846e732: UNAUTHORIZED: access to the requested resource is not authorized; map[]


2023-02-07T15:07:32.216Z        ERROR   Error during vulnerabilities scan: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
        * unable to inspect the image (quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:3ee77e3eb4702fbb1ea5dee009c2318405987c2fab352bdc14c67601515c4b6f): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
        * unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
        * containerd socket not found: /run/containerd/containerd.sock
        * GET https://quay.io/v2/openshift-release-dev/ocp-v4.0-art-dev/manifests/sha256:3ee77e3eb4702fbb1ea5dee009c2318405987c2fab352bdc14c67601515c4b6f: UNAUTHORIZED: access to the requested resource is not authorized; map[]

**Comment by admodev** [comment]
If you are running vulnerability scans on a private registry, you will have to log in to the docker account that has the images, or in this case, you will need to have the security rules on your openshift namespace to communicate with trivy image, like a proxy i would say...

**Comment by knqyf263** [comment]
>Trivy supports scanning images from [private registries](https://aquasecurity.github.io/trivy/v0.37/docs/advanced/private-registries/). Is it possible to make it work with quay.io? If yes, can we scan these openshift-* namespaces?

You can refer to [the doc](https://aquasecurity.github.io/trivy/v0.41/docs/advanced/private-registries/).

>Trivy takes 2-3 seconds before it could figure out that the image is not accessible. Are there any retries internally? Can we configure it not to spend much time on such images? From logs, it looks like trivy tries 4 ways to access the image (daemon, podman, contained and GET). Is it possible to skip first 3?

In progress: https://github.com/aquasecurity/trivy/issues/3049
